alphadia/transferlearning/train.py

```diff
@@ -927,51 +918,53 @@ def _test_ccs(
             Whether to continue training or not based on the early stopping criteria.
         """
         continue_training = True
-        if epoch % self.settings["test_interval"] == 0:
-            self.ccs_model.model.eval()
+        if epoch % self.settings["test_interval"] != 0 and epoch != -1:
```

Comment:
the four code blocks that are refactored here are very similar ..
we could try to further unify this, e.g. with https://refactoring.guru/design-patterns/template-method

Reply:
apparently Claude is with me ;-)

(just a proposal, needs further refinement I guess. I would e.g. inline `_prepare_test_input`)

1. Extract common logic in test methods:

```python
def _common_test_logic(self, model, test_df, metric_accumulator, epoch, epoch_loss, data_split, property_name):
    pred = model.predict(test_df)
    test_input = self._prepare_test_input(pred, test_df, property_name)
    val_metrics = metric_accumulator.calculate_test_metric(
        test_input, epoch, data_split=data_split, property_name=property_name
    )
    
    if epoch != -1 and data_split == "validation":
        self._accumulate_training_metrics(metric_accumulator, epoch, epoch_loss, model, property_name)
        val_loss = val_metrics[val_metrics["metric_name"] == "l1_loss"]["value"].values[0]
        continue_training = self.early_stopping.step(val_loss)
        self._log_training_progress(epoch, model, epoch_loss, val_loss)
    else:
        self._log_test_metrics(val_metrics, data_split)
    
    return continue_training

def _prepare_test_input(self, pred, test_df, property_name):
    if property_name == "ms2":
        return {
            "psm_df": test_df,
            "predicted": pred,
            "target": test_df,  # Assuming this is correct; adjust if needed
        }
    elif property_name == "charge":
        return {
            "target": np.array(test_df["charge_indicators"].values.tolist()),
            "predicted": np.array(pred["charge_probs"].values.tolist()),
        }
    else:  # rt and ccs
        return {
            "predicted": pred[f"{property_name}_pred"].values,
            "target": test_df[property_name].values,
        }

def _accumulate_training_metrics(self, metric_accumulator, epoch, epoch_loss, model, property_name):
    loss_name = "ce_loss" if property_name == "charge" else "l1_loss"
    metric_accumulator.accumulate_metrics(
        epoch,
        metric=epoch_loss,
        metric_name=loss_name,
        data_split="train",
        property_name=property_name,
    )
    current_lr = model.optimizer.param_groups[0]["lr"]
    metric_accumulator.accumulate_metrics(
        epoch,
        metric=current_lr,
        metric_name="lr",
        data_split="train",
        property_name=property_name,
    )

def _log_training_progress(self, epoch, model, epoch_loss, val_loss):
    current_lr = model.optimizer.param_groups[0]["lr"]
    logger.progress(
        f" Epoch {epoch:<3} Lr: {current_lr:.5f}   Training loss: {epoch_loss:.4f}   validation loss: {val_loss:.4f}"
    )

def _log_test_metrics(self, val_metrics, data_split):
    logger.progress(f" Model tested on {data_split} dataset with the following metrics:")
    for i in range(len(val_metrics)):
        logger.progress(
            f" {val_metrics['metric_name'].values[i]:<30}: {val_metrics['value'].values[i]:.4f}"
        )
```

2. Refactor test methods to use the common logic:

```python
def _test_ms2(self, epoch, epoch_loss, precursor_df, target_fragment_intensity_df, metric_accumulator, data_split, default_instrument="Lumos", default_nce=30.0):
    if epoch % self.settings["test_interval"] != 0 and epoch != -1:
        return True

    self.ms2_model.model.eval()
    precursor_df = precursor_df.copy()
    precursor_df["instrument"] = precursor_df.get("instrument", default_instrument)
    precursor_df["nce"] = precursor_df.get("nce", default_nce)

    continue_training = self._common_test_logic(
        self.ms2_model, precursor_df, metric_accumulator, epoch, epoch_loss, data_split, "ms2"
    )
    self.ms2_model.model.train()
    return continue_training

# Similar refactoring for _test_rt, _test_charge, and _test_ccs
```